src/mlpack/core/hpt/hpt.hpp

```diff
@@ -0,0 +1,268 @@
+/**
+ * @file hpt.hpp
+ * @author Kirill Mishchenko
+ *
+ * Hyper-parameter tuning.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_HPT_HPT_HPP
+#define MLPACK_CORE_HPT_HPT_HPP
+
+#include <mlpack/core/cv/meta_info_extractor.hpp>
+#include <mlpack/core/hpt/deduce_hp_types.hpp>
+#include <mlpack/core/optimizers/grid_search/grid_search.hpp>
+
+namespace mlpack {
+namespace hpt {
+
+/**
+ * The class HyperParameterTuner for the given MLAlgorithm utilizes the provided
+ * Optimizer to find the values of hyper-parameters that optimize the value of
+ * the given Metric. The value of the Metric is calculated by performing
+ * cross-validation with the provided cross-validation strategy.
+ *
+ * To construct a HyperParameterTuner object you need to pass the same arguments
+ * as for construction of an object of the given CV class. For example, we can
+ * use the following code to try to find a good lambda value for
+ * LinearRegression.
+ *
+ * @code
+ * // 100-point 5-dimensional random dataset.
+ * arma::mat data = arma::randu<arma::mat>(5, 100);
+ * // Noisy responses retrieved by a random linear transformation of data.
+ * arma::rowvec responses = arma::randu<arma::rowvec>(5) * data +
+ *     0.1 * arma::randn<arma::rowvec>(100);
+ *
+ * // Using 80% of data for training and remaining 20% for assessing MSE.
+ * double validationSize = 0.2;
+ * HyperParamterTuner<LinearRegression, MSE, SimpleCV> hpt(validationSize, data,
+ *     responses);
+ *
+ * // Finding the best value for lambda from the values 0.0, 0.001, 0.01, 0.1,
+ * // and 1.0.
+ * arma::vec lambdas{0.0, 0.001, 0.01, 0.1, 1.0};
+ * double bestLambda;
+ * std::tie(bestLambda) = hpt.Optimize(lambdas);
```

Comment:
It's not completely true. E.g. for `SoftmaxRegression` we have the following constructor:

```
  template<typename OptimizerType = mlpack::optimization::L_BFGS>
  SoftmaxRegression(const arma::mat& data,
                    const arma::Row<size_t>& labels,
                    const size_t numClasses,
                    const double lambda = 0.0001,
                    const bool fitIntercept = false,
                    OptimizerType optimizer = OptimizerType());
```
During hyper-parameter optimization we can only specify values for `lambda` and skip specifying values for `fitIntercept`. Then for `fitIntercept` the standard value (false) will be used.